pkg/scheduler/plugins/deviceshare/device_cache.go

```diff
@@ -0,0 +1,338 @@
+		}
+		if add {
+			n.allocateSet[deviceType][podNamespacedName] = struct{}{}
+		} else {
+			delete(n.allocateSet[deviceType], podNamespacedName)
+		}
+	}
+}
+
+func (n *nodeDevice) tryAllocateDevice(podRequest corev1.ResourceList) (apiext.DeviceAllocations, []error) {
+	allocateResult := make(apiext.DeviceAllocations)
+	var errors []error
+
+	for deviceType := range deviceResourceNames {
+		switch deviceType {
+		case schedulingv1alpha1.RDMA, schedulingv1alpha1.FPGA:
+			if !hasDeviceResource(podRequest, deviceType) {
+				break
+			}
+			if err := n.tryAllocateCommonDevice(podRequest, deviceType, allocateResult); err != nil {
+				errors = append(errors, err)
+			}
+		case schedulingv1alpha1.GPU:
+			if !hasDeviceResource(podRequest, deviceType) {
+				break
+			}
+			if err := n.tryAllocateGPU(podRequest, allocateResult); err != nil {
+				errors = append(errors, err)
+			}
+		default:
+			klog.Warningf("device type %v is not supported yet", deviceType)
+		}
+	}
+
+	return allocateResult, errors
+}
+
+func (n *nodeDevice) tryAllocateCommonDevice(podRequest corev1.ResourceList, deviceType schedulingv1alpha1.DeviceType, allocateResult apiext.DeviceAllocations) error {
+	podRequest = quotav1.Mask(podRequest, deviceResourceNames[deviceType])
+	nodeDeviceTotal := n.deviceTotal[deviceType]
+	if len(nodeDeviceTotal) <= 0 {
+		return fmt.Errorf("node does not have enough %v", deviceType)
+	}
+
+	var deviceAllocations []*apiext.DeviceAllocation
+
+	if isMultipleCommonDevicePod(podRequest, deviceType) {
+		var commonDeviceWanted int64
+		var podRequestPerCard corev1.ResourceList
+		switch deviceType {
+		case schedulingv1alpha1.RDMA:
+			commonDevice := podRequest[apiext.KoordRDMA]
+			commonDeviceWanted = commonDevice.Value() / 100
+			podRequestPerCard = corev1.ResourceList{
+				apiext.KoordRDMA: *resource.NewQuantity(commonDevice.Value()/commonDeviceWanted, resource.DecimalSI),
+			}
+		case schedulingv1alpha1.FPGA:
+			commonDevice := podRequest[apiext.KoordFPGA]
+			commonDeviceWanted = commonDevice.Value() / 100
+			podRequestPerCard = corev1.ResourceList{
+				apiext.KoordFPGA: *resource.NewQuantity(commonDevice.Value()/commonDeviceWanted, resource.DecimalSI),
+			}
+		}
+		satisfiedDeviceCount := 0
+		for minor, resources := range n.deviceFree[deviceType] {
+			if satisfied, _ := quotav1.LessThanOrEqual(podRequestPerCard, resources); satisfied {
+				satisfiedDeviceCount++
+				deviceAllocations = append(deviceAllocations, &apiext.DeviceAllocation{
+					Minor:     int32(minor),
+					Resources: podRequestPerCard,
+				})
+			}
+			if satisfiedDeviceCount == int(commonDeviceWanted) {
+				allocateResult[deviceType] = deviceAllocations
+				return nil
+			}
+		}
+		klog.V(5).Infof("node resource doest not satisfy pod's multiple %v request, expect %v, got %v", deviceType, commonDeviceWanted, satisfiedDeviceCount)
+		return fmt.Errorf("node does not have enough %v", deviceType)
+	}
+
+	for minor, resources := range n.deviceFree[deviceType] {
```

Comment:
else?

Reply:
Don't need an `else` here. If there is no resource matched pod request after for loop, we will return an error in the end.